packages/agents/src/mcp/do-oauth-client-provider.ts

```diff
@@ -124,17 +137,90 @@ export class DurableObjectOAuthClientProvider implements AgentsOAuthProvider {
     return this._authUrl_;
   }
 
-  /**
-   * Because this operates on the server side (but we need browser auth), we send this url back to the user
-   * and require user interact to initiate the redirect flow
-   */
+  stateKey(nonce: string) {
+    return `/${this.clientName}/${this.serverId}/state/${nonce}`;
+  }
+
+  async state(): Promise<string> {
+    const nonce = nanoid();
+    const state = `${nonce}.${this.serverId}`;
+    const storedState: StoredState = {
+      nonce,
+      serverId: this.serverId,
+      createdAt: Date.now()
+    };
+    await this.storage.put(this.stateKey(nonce), storedState);
```

Comment:
Minor, optional: I don't think this will happen, but wondering if we should clean these up lazily or via alarms in case many users end up leaving OAuth flows incomplete.